lib/um_arclight/package/generator.rb

```diff
@@ -0,0 +1,414 @@
+# frozen_string_literal: true
+
+require 'arclight'
+require 'benchmark'
+require 'json'
+require 'fileutils'
+require 'nokogiri'
+
+Deprecation.default_deprecation_behavior = :silence
+
+module UmArclight
+  module Package
+    # Generate HTML and PDF packages of finding aids
+    class Generator # rubocop:disable Metrics/ClassLength
+      COMPONENT_FIELDS = %w[
+        id
+        parent_ssi
+        parent_ssim
+        ref_ssi
+        ref_ssm
+        component_level_isim
+        normalized_title_ssm
+        level_ssm
+        scopecontent_teism
+        unitid_ssm
+        odd_tesim
+        bioghist_tesim
+        total_digital_object_count_isim
+        digital_objects_ssm
+        containers_ssim
+        repository_ssm
+      ].freeze
+
+      attr_accessor :identifier, :doc, :fragment, :collection, :session
+
+      def initialize(identifier:)
+        @identifier = identifier
+        @collection = nil
+        @session = ActionDispatch::Integration::Session.new(Rails.application)
+        @session.host = 'findingaids.lib.umich.edu'
+        @session.https!(true)
+      end
+
+      def build_html
+        response = get("/catalog/#{identifier}")
+        @doc = Nokogiri::HTML5(response.body)
+
+        components = []
+        elapsed_time = Benchmark.realtime do
+          @collection = fetch_doc(identifier)
+          components = fetch_components(identifier)
+        end
+        puts "UM-Arclight generate package : #{collection.id} : fetch components (in #{elapsed_time.round(3)} secs)."
+        elapsed_time = Benchmark.realtime do
+          @fragment = render_fragment(
+            collection: collection,
+            components: components
+          )
+
+          update_navigation_links
+          update_package_html
+        end
+        puts "UM-Arclight generate package : #{collection.id} : build HTML (in #{elapsed_time.round(3)} secs)."
+      end
+
+      def generate_html
+        build_html
+
+        output_filename = generate_output_filename('.html')
+        FileUtils.makedirs(File.dirname(output_filename)) unless Dir.exist?(File.dirname(output_filename))
+
+        File.open(output_filename, 'w') do |f|
+          f.puts doc.serialize
+        end
+      end
+
+      def build_pdf
+        # build the source in tmp
+        FileUtils.mkdir_p(working_path_name)
+        Dir.chdir(working_path_name)
```

Comment:
I guess the other pieces are dependent on cwd... We'll probably want to leak test this for threading issues, though it should always be running in a single-threaded context. There might be some potential for doing this in a block, like some of the tempfile and popen3 constructs, to make the "working dir-ness" of this whole operation more explicit.